### PR TITLE
Support callback subclasses in `configure_callbacks`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Passing a subclass of one of the Trainer default callbacks would not replace it.

Part of #16990

cc @awaelchli @justusschock @borda

### CODE REVIEW

File: src/lightning/pytorch/trainer/connectors/callback_connector.py [hunk 1]
```diff
@@ -172,7 +172,15 @@ def _attach_model_callbacks(self) -> None:
...
-        override_types = model_callback_types.intersection(trainer_callback_types)
+        # edge case: if an unmodified callback was added, the logic below would filter it
+        trainer_callback_types.discard(Callback)
+        # exclude trainer callbacks of the same class or subclass
+        override_types = set()
+        for m in model_callback_types:
+            for c in trainer_callback_types:
+                if issubclass(m, c):
+                    override_types.add(c)
+                    break
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/callback_connector.py:183]: This break assumes nobody is doing multiple inheritance with our existing callbacks

```python
class Frankenstein(ModelCheckpoint, TQDMProgressBar):
    ...
```

Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/callback_connector.py:183]: This PR fixes  #16990 if I understand correctly. I added it to the description.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/trainer/connectors/callback_connector.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/connectors/callback_connector.py:183]: Oh I didn't know about that issue. There you describe more edge cases that we could want to consider, so maybe it's worth keeping open

File: src/lightning/pytorch/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -172,7 +172,15 @@ def _attach_model_callbacks(self) -> None:
...
         trainer_callback_types = {type(c) for c in trainer.callbacks}
-        override_types = model_callback_types.intersection(trainer_callback_types)
+        # edge case: if an unmodified callback was added, the logic below would filter it
+        trainer_callback_types.discard(Callback)
+        # exclude trainer callbacks of the same class or subclass
+        override_types = set()
+        for m in model_callback_types:
+            for c in trainer_callback_types:
+                if issubclass(m, c):
+                    override_types.add(c)
```
Comment (Borda) [inline src/lightning/pytorch/trainer/connectors/callback_connector.py:182]:
```suggestion
        for m_cb in model_callback_types:
            for t_cb in trainer_callback_types:
                if issubclass(m_cb, t_cb):
                    override_types.add(t_cb)
```

**Review by Borda — APPROVED** [review]

**Pull request merged**